wdl-analysis/src/types/v1.rs

```diff
@@ -657,31 +657,32 @@ impl<'a, C: EvaluationContext> ExprTypeEvaluator<'a, C> {
     /// Checks a placeholder expression.
     pub(crate) fn check_placeholder<N: TreeNode>(&mut self, placeholder: &Placeholder<N>) {
         self.placeholders += 1;
-
         // Evaluate the placeholder expression and check that the resulting type is
         // coercible to string for interpolation
         let expr = placeholder.expr();
         if let Some(ty) = self.evaluate_expr(&expr) {
-            match ty {
-                Type::Primitive(..) | Type::Union | Type::None => {
-                    // OK
-                }
-                ty => {
-                    // Check for a sep option is specified; if so, accept `Array[P]` where `P` is
-                    // primitive.
-                    let mut coercible = false;
-                    if let Some(PlaceholderOption::Sep(_)) = placeholder.option() {
-                        if let Type::Compound(CompoundType::Array(ty), _) = &ty {
-                            if !ty.element_type().is_optional()
-                                && ty.element_type().as_primitive().is_some()
-                            {
-                                // OK
-                                coercible = true;
-                            }
-                        }
+            if let Some(option) = placeholder.option() {
+                let valid = match option {
+                    PlaceholderOption::Sep(_) => matches!(ty,
+                        Type::Compound(CompoundType::Array(ref element_type), false)
+                        if matches!(element_type.element_type(), Type::Primitive(_, false) | Type::Union)),
+                    PlaceholderOption::Default(_) => matches!(ty, Type::Primitive(_, true)),
+                    PlaceholderOption::TrueFalse(_) => {
+                        matches!(ty, Type::Primitive(PrimitiveType::Boolean, _))
```

Comment:
As `true/false` expects a non-optional `Boolean`, we should also be accepting only `Union` here.
 
```suggestion
                        matches!(ty, Type::Primitive(PrimitiveType::Boolean, false) | Type::Union)
```